Put labels code back in

ptersilie opened the pull request:
There we go. Let's get this merged into your fork and then we can raise a proper PR against softdevteam from there.

ptersilie committed:
Put labels code back in

vext01 commented on src/librustc_codegen_ssa/mir/block.rs:
```diff
@@ -776,6 +777,8 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
 
         debug!("codegen_block({:?}={:?})", bb, data);
 
+        let llbb = bx.llbb();
```
shall we move that closer to where it is used?

ptersilie replied:
We can't. I put this there because I'm worried it will be overwritten by any of the statements. That being said, it seems way back we also never passed the block in (and instead used `llbb()` in `add_yk_block_label`, which just gets the current block the builder is pointing to). Since we never had any problems with this, it might mean that the statements don't overwrite this. But it still makes me nervous that we can't be sure.

vext01 replied:
Good point. Shall we add a comment?

vext01 commented on src/librustc_codegen_ssa/mir/block.rs:
```diff
@@ -784,17 +787,19 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
-        //    bx.add_yk_block_label(lbl_name);
-        //}
+        if bx.tcx().sess.opts.cg.tracer.sir_labels() &&
+            !bx.tcx().def_path_str(self.instance.def_id()).contains("drop_in_place") &&
```
I think these should be exact equality checks so as to not skip user functions that may contain "drop_in_place".

IIRC, the two offenders are `core::drop_in_place` and `std::drop_in_place`, but might be wrong.

vext01 commented on src/librustc_codegen_ssa/mir/block.rs:
```diff
@@ -784,17 +787,19 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
+            !bx.tcx().crate_name(LOCAL_CRATE).as_str().starts_with("rustc") {
+            use ykpack::BLOCK_LABEL_PREFIX;
+            let lbl_name = CString::new(format!(
+                    "NEW_{}:{}:{}",
```
Oops, I should have killed the `NEW_` prefix. Let's kill it now.

vext01 replied:
Didn't we have a constant defined someplace for this in ykpack?

vext01 replied:
Ah yes. We use it! Sorry.

So just kill `NEW_`

vext01 commented:
Just a few comments!

ptersilie commented:
Addressed your comments. Although there are other `drop_in_place` functions, it seems we get away with only skipping those three.

vext01 commented:
LGTM.

You can squash if you like, but we will probably rebase the target branch later anyway.

Up to you.

ptersilie commented:
Squashed.

ptersilie commented:
Actually, if you want to wait a minute I can add another commit that adds the function return labels too.

vext01 commented:
ok

vext01 commented:
Is this ready for re-review?

ptersilie commented:
I've just added the labels for function call returns. I've also deduplicated some code and moved it into `BuilderCx`. I would have liked to just use `interface` as a reference and forward that, but the compiler wouldn't let me, so we are now fowarding the symbol and the function names.

ptersilie commented:
Yes, this is ready for re-review.

vext01 commented on src/librustc_codegen_ssa/mir/block.rs:
```diff
@@ -141,6 +141,22 @@ impl<'a, 'tcx> TerminatorCodegenHelper<'tcx> {
+                        bx.cx().tcx().symbol_name(fx.instance).name.as_str() != "main" &&
+                        !bx.tcx().crate_name(LOCAL_CRATE).as_str().starts_with("rustc") {
+                            let llbb = bx.llbb();
+                            use ykpack::BLOCK_LABEL_PREFIX;
```
Do we need another prefix for this kind of label?

ptersilie replied:
I guess we could but it's not strictly necessary.

vext01 replied:
If a block contains a call to a function with a return value, won't we insert two labels of the same name at different places?

I may be wrong, I don't 100% remember the details.

ptersilie replied:
My gut feeling says no, but I'm not sure. Let's rename it just in case.

ptersilie replied:
Ah damn. Because I moved the label generation into `BuilderCx` I now need to add another argument to the `add_yk_block_label` function.

vext01 commented:
> I would have liked to just use interface as a reference and forward that, but the compiler wouldn't let me

You meant the `instance`, right? I'm OK with passing the symbol name and index.

ptersilie commented:
Sorry, yes, I meant the `instance`.

ptersilie commented:
Moved the label generation code to after `codegen_terminators` which means we are now getting labels for all mir blocks (as far as I can tell). This means we can removed the label generation code for function returns. Ready to merge into your branch.

vext01 commented:
bors r+

vext01 commented:
Let's merge this manually, as bors will test this when my branch is merged.
